// tcomb/lib/maybe.js
var assert = require('tcomb/lib/assert');
var isTypeName = require('tcomb/lib/isTypeName');
var isFunction = require('tcomb/lib/isFunction');
var isMaybe = require('tcomb/lib/isMaybe');
var isIdentity = require('tcomb/lib/isIdentity');
var Any = require('tcomb/lib/Any');
var create = require('tcomb/lib/create');
var Nil = require('tcomb/lib/Nil');
var forbidNewOperator = require('tcomb/lib/forbidNewOperator');
var is = require('tcomb/lib/is');
var getTypeName = require('tcomb/lib/getTypeName');

function getDefaultName(type) {
  return '?' + getTypeName(type);
}

function maybe(type, name) {

  if (isMaybe(type) || type === Any || type === Nil) { // makes the combinator idempotent and handle Any, Nil
    return type;
  }

  if (process.env.NODE_ENV !== 'production') {
    assert(isFunction(type), function () { return 'Invalid argument type ' + assert.stringify(type) + ' supplied to maybe(type, [name]) combinator (expected a type)'; });
    assert(isTypeName(name), function () { return 'Invalid argument name ' + assert.stringify(name) + ' supplied to maybe(type, [name]) combinator (expected a string)'; });
  }

  var displayName = name || getDefaultName(type);
  var identity = isIdentity(type);

  function Maybe(value, path) {
    if (process.env.NODE_ENV !== 'production') {
      if (identity) {
        forbidNewOperator(this, Maybe);
      }
    }
    return Nil.is(value) ? value : create(type, value, path);
  }

  Maybe.meta = {
    kind: 'maybe',
    type: type,
    name: name,
    identity: identity
  };

  Maybe.displayName = displayName;

  Maybe.is = function (x) {
    return Nil.is(x) || is(x, type);
  };

  return Maybe;
}

maybe.getDefaultName = getDefaultName;
module.exports = maybe;
